README.md
# Scalaser

## Overview

The scalaser package was created as a part of the bachelor thesis of Miro Voellmy. It was specifically developed for stair climbing wheelchair Scalevo.

The scalaser package has been tested under [ROS] Indigo and Ubuntu 14.04. This is research code, expect that it changes often and any fitness for a particular purpose is disclaimed.

The source code is released under a [BSD 3-Clause license](ros_package_template/LICENSE).

**Author: Miro Voellmy  
Contact: Miro Voellmy, mvoellmy@ethz.ch  
Affiliation: Autonomous Systems Lab, ETH Zurich**

![Example image](doc/scalebot_on_stairs.jpg)


### Publications

If you use this work in an academic context, please cite the following publication(s):
	
* M. Voellmy, P. Fankhauser, M. Hutter, and R. Siegwart: **Stairrecognition on a stair climbing wheelchair using LIDAR Sensors**. IEEE/RSJ International Conference of Placeholders (IROS), 2015. ([PDF](https://www.youtube.com/watch?v=dQw4w9WgXcQ))


## Installation

### Dependencies

- [Robot Operating System (ROS)](http://wiki.ros.org) (middleware for robotics),
- [Eigen](http://eigen.tuxfamily.org) (linear algebra library),
- [Matlab](http://ch.mathworks.com/products/matlab/) (matlab),
- [c_matlabcppinterface](https://bitbucket.org/adrlab/c_matlabcppinterface/wiki/Home) (matlab C++ interface),
- [scalevo_msgs](http://www.github.com/scalevo/scalevo_msgs) (messages package of scalevo),
- [scalebot](http://www.github.com/scalevo/scalebot) (robot model of the scalevo wheelchair)
### Building

In order to install, clone the latest version from this repository into your catkin workspace and compile the package using

	cd catkin_workspace/src
	git clone https://github.com/scalevo/scalaser.git
	cd ../
	catkin_make

### Tests

Package tests have not been implemented yet. You can run this command however to watch Starwars Episode IV in the terminal: 

	telnet towel.blinkenlights.nl


## Usage

First start the two SICK sensors on board of the Scalevo wheelchair using

	roslaunch scalaser pointcloud_2.launch

then start the stair recognition node with

	roslaunch scalaser angle.launch

Check the launch files if you want to change a modifiable parameters.

## Nodes

### Angle

Main node to determine the angle of the stair on stairs.

#### Subscribed Topics

* **`cloud_1`** ([sensor_msgs/PointCloud])

	Pointcloud of the right laser sensor.
	
* **`cloud_2`** ([sensor_msgs/PointCloud])

	Pointcloud of the left laser sensor.


#### Published Topics

* **`beta`** ([std_msgs/Float64])

	Angle of the wheelchair on the stairs. 

* **`stair_parameters`** ([std_msgs/Float64MultiArray])
	
	Determined parameters of the stair and the wheelchairs position on the stair.

* **`stair_model`** ([visualization_msgs/Marker])

	The marker for the stair visualization in [rviz].

* **`chair_model`** ([visualization_msgs/Marker])

	The marker for the chair visualization in [rviz].

#### Services

* **`trigger`** ([std_srvs/Empty])

	Trigger the computation process. For example, you can trigger the computation from the console with

		rosservice call /NODE_A_NAME/trigger


#### Parameters

* **`scalaser/fov_s`** (int, default: 200)

	The start of the field of view.
		
* **`scalaser/fov_s`** (int, default: 150)

	The size of the field of view.

* **`scalaser/kp`** (double, default: 0.008)
	
	P value used to determine the motor velocity.


### StairModel

Publishes a model of the stairs for visualization in rviz.

### ChairModel

Publishes a model of the wheelchair positioned on the stairs for visualization in rviz.


## Bugs & Feature Requests

Please report bugs and request features using the issue feature of this package.

[ROS]: http://www.ros.org
[rviz]: http://wiki.ros.org/rviz
[Eigen]: http://eigen.tuxfamily.org
[Gazebo]: http://gazebosim.org/
[rviz]: http://wiki.ros.org/rviz
[starleth_msgs/SeActuatorCommands]: https://bitbucket.org/ethz-asl-lr/c_starleth_ros_common/raw/master/starleth_msgs/msg/SeActuatorCommands.msg
[grid_map_msg/GridMap]: https://github.com/ethz-asl/grid_map/blob/master/grid_map_msg/msg/GridMap.msg
[sensor_msgs/PointCloud2]: http://docs.ros.org/api/sensor_msgs/html/msg/PointCloud2.html
[visualization_msgs/Marker]: http://docs.ros.org/api/visualization_msgs/html/msg/Marker.html
[nav_msgs/OccupancyGrid]: http://docs.ros.org/api/nav_msgs/html/msg/OccupancyGrid.html
[std_srvs/Empty]: http://docs.ros.org/api/std_srvs/html/srv/Empty.html


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package scalaser
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

TODO
------------------
* add kp to plot
* change matlab plots to change data setXdata, setYdata
* define beta threshold in launch file
* add folder for figures

KNOWN BUGS
------------------
* Fix inconsistent performance of the angle node. It ranges from 1 Hz to 4 Hz. If a rviz is started the frequency takes a hit, however once it's closed again the node wont performe as well as before.
* joint_states influence towards the initial guesses causes a big mess for the stair_model


NICE TO HAVE
------------------
* add radius-tracking
* add 3D model for curved stairs
* field of view initialization
* add field of view in rviz
* clean up v_r_1 and dx_1 in angle.cpp
* add initialization for start and end of stairs to determine when to stop the stair model
* add timestamp to published messages
* make fov_s, fov_d dependant of the lambda position received from the MyRIO
* make stair growth dependant of fov_s
* make the velocity_forward sent to the MyRio adjustable with dynamic reconfigure
* (use IMU values to set guess for beta threshold value -> not sure if needed, since the wheelchair never comes close to a beta=~24°)
* (change the initialization of v0 from within the angle constructor to the matching constructor since startvalues for both sides are identical)
* (change tf static to tf2 static)

0.1.00 (2015-06-12)
-------------------
* if fmincon did not converge the angle computations are not executed, additionally if this happens for more than 3 times matching gets reinitialized
* commented everything within joint_callback out
* stair model size has been updated
* changed dx reinitialization to only happen if the stair is acctually curved 

0.0.14 (2015-06-10)
-------------------
* added filter for scan data to be between 0.3 and 1 m
* changed beta threshold

0.0.14 (2015-06-08)
-------------------
* fixed incorrect alpha computation --> hurray :D
* implemented interpolation between edges to get a constant alignment over the stairs

0.0.14 (2015-06-08)
-------------------
* changed matchTemplate() from void to VectorXd in order to directly copy v_r to angle.cpp
* added marker which displays the position of the edge around the wheelchair orients itself
* fixed a bug which caused beta to change sign while going up or down stairs
* changed the published vel_rot to be negative to compensate for the negative z-axis in LabView
* beta plot gets saved with current timestamp to the matlab folder
* now all scanpoints which are closer than a threshold get discarded
* info message is added which indicates the switch from one edge to another
* the data from plot_data() gets continiously plotted after 20 timer callbacks

0.0.13 (2015-06-05)
-------------------
* Added first draft of Double Edge algorithm
* Now outputs the count of NULL values within FoV
* Reinitializing updates values from parameter server correctly now.
* Implemented better FoV reduction.
* Now subs scan_1/2 instead of cloud_1/2 and does pointcloud transform itself. First the full scan gets added to a vector. Within setData() it then gets reduced and transformed.
* Dzi and phi0 guesses are received from the myRio.

0.0.12 (2015-06-04)
-------------------
* changed FoV reduction to be computed out of the pointcloud points.
* beta filter has been changed to only publish a beta if it is between +-10°

0.0.11 (2015-06-03)
-------------------
* added parameter vel_fwd which is published to set_vel as a way to give the motors speed while testing
* changed the beta filter to reinitialize matching if beta is bigger than +-10°
* readded the if-condition again to check if the rotation velocity can be published.

0.0.10 (2015-06-02)
-------------------
* fixed a bug, which caused the stair tf to get all messed up
* realized, that the pointcloud vector isn't always of size 811 and thus the method of reduzing its fov doesn't work that robust

0.0.9 (2015-05-19)
------------------
* created launch-file to start the wheelchair alignment for testing
* updated README
* added tests section to changelog

0.0.8 (2015-05-18)
------------------
* robot model of the wheelchair has been created in the package SCALEBOT 
* dzi, phi0 are now computed out of the lambda positions received from the MyRIO -> this might only be needed during initialization to save computation time.
* if beta values are unusual for a number of computations, matching gets reinitialized
* fov_s & fov_d are editable before restart of the service

0.0.7 (2015-05-15)
------------------
* changed the velocity publisher from Float64MultiArray to String
* now plots beta after service is finished
* replaced all 3.14 values with PI
* service now only starts the initialization if pointclouds are published

0.0.6 (2015-05-13)
------------------
* resets beta_old and beta_new before restarting service
* added complete wheelchair model to visualtization
* created launch-file to start the visualization (stair_viz.launch)
* added rviz file
* stair visualization now grows dependant on distance traveled on the stairs

0.0.5 (2015-05-12)
------------------
* now sets boundary constrains of dx to "dx_old ± (stair-diagonal)/2" to stop phase offset completely
* changed initializer vector v0 from <vector> to Eigen
* now initializes v0 and boundry contraints with the start of the ros service. The result vector of the first matching are used as the start values for the first matching of the other side.
* found error that causes beta to randomly jump ~20° caused by phase offset of dx_1 and dx_2

  Approach:
  - solution_1: tracking of the motor encoders helps to determine if dx makes sense and changes dx accordingly
  - solution_2: using the fminsearch results from the other side as start values reduces chances of dx beeing phaseoffset
  - solution_3: filter which checks previous dx value and beta value and compares them to check if the difference make any sense

src/angle.cpp
#include "angle.h"

Angle::Angle(ros::NodeHandle n_):
n(n_), a(.7), v_s(5), beta_old(0), beta_new(0), v0(5), count(0), wrong_beta_count(0), wrong_se_r_count(0)
r_h(.1016), s(.653837), phi_f(.193897),
kp(0), vel_fwd(0),
v_r_1(2), v_r_2(2),
alpha(0)
{
  setParameters();
  initializePlotEngine();
  initializeEdge();


  v0 << 0.17, 0.3, 0.0, 0.0, 0.0;

  // ROS_INFO("Start Values for fminsearch:");
  // ROS_INFO("stair heigth________h0 = %f",v0(0); 
  // ROS_INFO("stair depth_________t0 = %f",v0(1);
  // ROS_INFO("phase offset_______dx0 = %f",v0(2);
  // ROS_INFO("sensor height______dz0 = %f",v0(3);
  // ROS_INFO("sensor rotation___phi0 = %f",v0(4);

  // create & asssign temporary matching object
  matching cloud_1_t(n_,phi0,dzi,fov_s,fov_d,v0,1);
  matching cloud_2_t(n_,phi0,dzi,811-fov_s-fov_d,fov_d,v0,2);
  cloud_1 = cloud_1_t;
  cloud_2 = cloud_2_t;

  // initialize main timer
  main_timer = n.createTimer(ros::Duration(0.2),&Angle::timerCallback,this,false,false);

  // initialize service
  service = n.advertiseService("align_wheelchair",&Angle::alignWheelchair,this);

  ROS_INFO("Node started. Waiting for input.");

  ros::spin();
}

bool Angle::alignWheelchair(scalevo_msgs::Starter::Request& request, scalevo_msgs::Starter::Response& response) {
  if (request.on) {

    // create subscriber
    sub_1 = n.subscribe("scan_1", 1, &matching::matchCallback, &cloud_1);
    sub_2 = n.subscribe("scan_2", 1, &matching::matchCallback, &cloud_2);

    sub_joint = n.subscribe("/joint_states", 1, &Angle::jointCallback, this);


    // advertise topics
    pub_1 = n.advertise<std_msgs::Float64>("/beta", 1);
    pub_2 = n.advertise<std_msgs::Float64MultiArray>("/stair_parameters", 1);
    // pub_velocity = n.advertise<std_msgs::Float64MultiArray>("/set_vel", 100);
    pub_s_velocity = n.advertise<std_msgs::String>("/scalevo_cmd", 1);
    pub_edge_marker = n.advertise<visualization_msgs::Marker>("edge_marker", 1);

    while (sub_1.getNumPublishers() == 0 || sub_2.getNumPublishers() == 0) {
      sleep(1);
    }

    ros::spinOnce();
    
    initializeMatching();
    
    // start main loop timer
    main_timer.start();

    ROS_INFO("Wheelchair alignment has been started.");
    return true;  

  }

  else if (!request.on) {
    // shutdown subscribers
    sub_1.shutdown();
    sub_2.shutdown();
    sub_joint.shutdown();
    
    // shutdown publishers
    pub_1.shutdown();
    pub_2.shutdown();
    // pub_velocity.shutdown();
    pub_s_velocity.shutdown();
    pub_edge_marker.shutdown();

    // stop main loop timer
    main_timer.stop();

    ROS_INFO("Nr of computations: %d", count);
    ROS_INFO("Duration:           %f", ros::Time::now().toSec() - time_start);
    ROS_INFO("Average frequency:  %f Hz",count/(ros::Time::now().toSec() - time_start));
    ROS_INFO("Wheelchair alignment has been stopped.");

    plotData(beta_vector);
    plot_engine.executeCommand("savefig(datestr(now))");
    ROS_INFO("Plot has been saved to file.");

    return true;
  }
  return true;
}

void Angle::initializePlotEngine() {
  // plot data in a new matlab engine
  if (plot_engine.initialize() && plot_engine.good()) ROS_INFO("Plot engine succesfully initialized.");
  plot_engine.changeWorkingDirectory("~/catkin_ws/src/scalaser/matlab");
  // plot_engine.showWorkspace();
  plotData(beta_vector);

}

void Angle::timerCallback(const ros::TimerEvent& event) {
  if (sub_1.getNumPublishers() > 0 && sub_2.getNumPublishers() > 0) {

    cloud_1.setData();
    cloud_2.setData();
    v_r_1 = cloud_1.matchTemplate();
    v_r_2 = cloud_2.matchTemplate();

    if (cloud_1.getSe_r() < threshold && cloud_2.getSe_r() < threshold) {
      computeAngle();
      computeStair();
      pubStairParameters();
      pubEdge();
    }
    else {
      ROS_WARN("Nothing published because fmincon did not converge.");
    }

    computeVelocity();

    ROS_INFO("Callback time:        %f",event.profile.last_duration.toSec());
    count++;
  }

 //  if (fmod(count,5) == 0) {
 //    plotData(beta_vector);
 // }
}

void Angle::jointCallback(const sensor_msgs::JointState::ConstPtr& joint_state) {
  // phi0 = -joint_state -> position[0];
  // dzi = r_h + s*sin(-phi0 + phi_f);

  // to only set parameters after reinitialization comment this code and write phi0 and dzi to the parameter server instead
  // cloud_1.setParameters(phi0, dzi, fov_s, ofv_d);
  // cloud_2.setParameters(phi0, dzi, 811-fov_s-fov_d, fov_d);
  // ros::param::set("/scalaser/phi", PI/180*phi0);
  // ros::param::set("/scalaser/dzi", dzi);

  // ROS_INFO("Matching Parameters have been updated.");
}

void Angle::initializeMatching() {

  setParameters();

  beta_new = 0;
  beta_old = 0;
  beta_vector.clear();
  alpha_vector.clear();
  dx_1_vector.clear();
  dx_2_vector.clear();
  time_vector.clear();

  count = 0;

  cloud_1.setParameters(phi0,dzi,fov_s,fov_d);
  cloud_2.setParameters(phi0,dzi,811-fov_s-fov_d,fov_d);


  ROS_INFO("Start time: %f",time_start);
  
  setBoundaries();

  time_start = ros::Time::now().toSec();
  count++;
}

void Angle::setBoundaries() {
  cloud_1.setData();
  cloud_2.setData();
  cloud_1.setFminArgs(v0);
  v_r_1 = cloud_1.matchTemplate();
  cloud_2.setFminArgs(v_r_1);
  v_r_2 = cloud_2.matchTemplate();
}

void Angle::computeAngle() {

  dx_1 = cloud_1.getDx();
  dx_2 = cloud_2.getDx();
  diag_1 = cloud_1.getDiag();
  diag_2 = cloud_2.getDiag();

// Experimental for Curve
  if (fabs(alpha) > 1) {
    if (fabs(dx_1) > 0.8*diag_1 || fabs(dx_2) > 0.8*diag_2) {
      v0 = cloud_1.getV_r();
      v0(2) = 0;
      setBoundaries();
      dx_1 = cloud_1.getDx();
      dx_2 = cloud_2.getDx();
      diag_1 = cloud_1.getDiag();
      diag_2 = cloud_2.getDiag();
      ROS_INFO("------Edge has been changed!------");
    }
  }
  computeAlpha();
  // updateFoV();
  // setFoV();
  // setDx();
// Endperimental

  computeBeta();
}

void Angle::computeAlpha() {
  // Get Parameters from matching objects
  alpha_1 = atan((dx_2 - dx_1) / a);
  alpha_2 = atan(((diag_2 - dx_2) - (diag_1 - dx_1)) / a);
  alpha = 180/PI*(alpha_1 + alpha_2);
  
  // ROS_INFO("ALPHA_1: %f°", 180/PI*(alpha_1));
  // ROS_INFO("ALPHA_2: %f°", 180/PI*(alpha_2));
  // ROS_INFO("ALPHA: %f°", alpha);
  // ROS_INFO("BETAA: %f°", beta_new);
  // ROS_INFO("BETAB: %f°", 180/PI*(alpha_1));
}

void Angle::computeBeta() {

  // beta_new = 180/PI*atan((dx_2 - dx_1)/a); // Single Edge Beta Computation

  beta_new = 180/PI*(alpha_1 - alpha_2)/2;
  ROS_INFO("ALPHA:         %f°", alpha);
  ROS_INFO("BETA_SIMPLE:   %f°", beta_new);

    // Linear interpolation between two edges
  d_beta = alpha*(dx_2 + dx_1)/(diag_2 + diag_1) - alpha/2;
  beta_new -= d_beta; // + or - not sure yet

  ROS_INFO("D_BETA:        %f°", d_beta);
  ROS_INFO("BETA_INTERP:   %f°", beta_new);

  // if (fabs(beta_old - beta_new) < 15 && fabs(beta_old) < 10) {
  if (fabs(beta_new) < 15) {

    beta.data = beta_new;
    pub_1.publish(beta);

    ROS_INFO("Time: %f",ros::Time::now().toSec()-time_start);
    time_vector.push_back(ros::Time::now().toSec()-time_start);
    alpha_vector.push_back(180/PI*(alpha_1 + alpha_2));
    beta_vector.push_back(beta_new);

    v0 = cloud_1.getV_r();
    dx_1_vector.push_back(v0(2));
    v0 = cloud_2.getV_r();
    dx_2_vector.push_back(v0(2));

    wrong_beta_count = 0;
  }
  else {
    wrong_beta_count++;
    ROS_WARN("No beta published since unreasonable values occured. ⁻ %d",wrong_beta_count);
    ROS_WARN("Refused beta: %f",beta_new);
    if (wrong_beta_count > 3) initializeMatching();
  }
  beta_old = beta.data;
}

void Angle::computeStair() {
  v_s = (cloud_1.getV_r()+cloud_2.getV_r())/2*cos(beta.data*PI/180);
  stair_param.data.clear();
  for (int i=0; i < 5; i++)  stair_param.data.push_back(v_s(i));
  pub_2.publish(stair_param);
}

void Angle::computeVelocity() {

  // Read Velocity Parameters from Server
  n.param("/scalaser/kp",kp,0.05);
  n.param("/scalaser/vel_fwd",vel_fwd,0.0);

  // velocity.data.clear();
  // velocity.data.push_back(0);
  // if(cloud_1.getSe_r()<threshold && cloud_2.getSe_r()<threshold)
  //   {velocity.data.push_back(-beta.data*PI/180*kp);}
  // else {
  //   velocity.data.push_back(0);
  //   ROS_WARN("No velocity published since matching didn't work properly.");
  // }

  // Fill velo message
  std_msgs::String velo;

  velo.data = "set_vel,"; 
  std::ostringstream buff_1;
  buff_1 << vel_fwd;
  velo.data += buff_1.str();
  velo.data += ","; 
  std::ostringstream buff_2;

  if(cloud_1.getSe_r() < threshold && cloud_2.getSe_r() < threshold) {
    buff_2 << - beta.data * kp;
    velo.data += buff_2.str();
    wrong_se_r_count = 0;
  }
  else {
    wrong_se_r_count++;
    buff_2 << 0;
    velo.data += buff_2.str();
    if (wrong_se_r_count > 3) initializeMatching();
  }
  pub_s_velocity.publish(velo);
}

void Angle::pubStairParameters() {

  double h = v_s(0);
  double t = v_s(1);
  double dx = -v_s(2);
  double dz = dzi + v_s(3);
  double phi = -phi0 - v_s(4);

  transform.setOrigin( tf::Vector3(-(cos(phi)*dx + sin(phi)*dz),0,-(-sin(phi)*dx + cos(phi)*dz)) );
  tf::Quaternion q;
  q.setRPY(0,phi,-beta.data*PI/180);
  transform.setRotation(q);
  br.sendTransform(tf::StampedTransform(transform,ros::Time::now(),"laser_mount_link","stair_middle"));
  ROS_INFO_ONCE("Transform to stair sent");
}

void Angle::setParameters() {
  n.param("/scalaser/fov_s",fov_s,200);
  n.param("/scalaser/fov_d",fov_d,150);
  n.param("/scalaser/dzi",dzi,.65);
  n.param("/scalaser/phi",phi0,-43.0);
  n.param("/scalaser/kp",kp,0.05);
  n.param("/scalaser/vel_fwd",vel_fwd,0.0);
  n.param("/scalaser/threshold",threshold,0.08);

  phi0 = phi0*PI/180;

  ROS_INFO("Phi0: %f",phi0);
  ROS_INFO("Field of View Start: %d",fov_s);
  ROS_INFO("Field of View Window: %d",fov_d);
  ROS_INFO("Dz Initial: %f",dzi);

  ROS_INFO("Parameters from Server have been updated.");
}

void Angle::plotData() {
  plot_engine.executeCommand("clf");
  plotData(beta_vector);
  plotData(alpha_vector);
  // plotData(dx_1_vector);
  // plotData(dx_2_vector);

  plot_engine.executeCommand("xlabel('Time [s]')");
  plot_engine.executeCommand("legend('beta [°]', 'alpha [°]', 'step diagonal delta [m]')");
}

void Angle::plotData(std::vector<double> data_vector) {
  Eigen::VectorXd data_vectorXd(data_vector.size());
  Eigen::VectorXd time_vectorXd(time_vector.size());

  for (int i=0; i < data_vector.size(); ++i) data_vectorXd[i] = data_vector[i];
  for (int i=0; i < time_vector.size(); ++i) time_vectorXd[i] = time_vector[i];

  plot_engine.put("data_vector", data_vectorXd);
  plot_engine.put("time_vector",time_vectorXd);

  plot_engine.executeCommand("plot_data(time_vector, data_vector)");


  ROS_INFO("Results have been plotted.");

  data_vector.clear();

  // To save a std::vector into a .txt file use:
    // std::ofstream file_beta;
    // file_beta.open("/home/miro/Desktop/cpp2matlab/beta_vector.txt");
    // for(std::size_t i = 0; i < beta_vector.size(); ++i) file_beta << beta_vector[i] << std::endl;
    // file_beta.close();

    // std::ofstream file_time;
    // file_time.open("/home/miro/Desktop/cpp2matlab/time_vector.txt");
    // for(std::size_t i = 0; i < time_vector.size(); ++i) file_time << time_vector[i] << std::endl;
    // file_time.close();
}

void Angle::initializeEdge() {
  edge_marker.header.frame_id = "laser_mount_link";
  edge_marker.ns = "edge";
  edge_marker.id = 1;
  edge_marker.type = visualization_msgs::Marker::ARROW;
  edge_marker.action = 0;
  //edge_marker.pose.orientation.w = 1.0;
  edge_marker.scale.x = 0.05;
  edge_marker.scale.y = 0.05;
  edge_marker.scale.z = 0.001;
  edge_marker.color.a = 1.0; // Don't forget to set the alpha!
  edge_marker.color.r = 0.0;
  edge_marker.color.g = 1.0;
  edge_marker.color.b = 0.0;
}

void Angle::pubEdge() {
  edge_marker.header.stamp = ros::Time::now();
  // Publish Edge around which the wheelchair orients itf
  geometry_msgs::Point p;

  p.x = -( cos(- phi0 - v_r_1(4))*-v_r_1(2) + sin(- phi0 - v_r_1(4))*(dzi + v_r_1(3)));
  p.y = -a/2;
  p.z = -(-sin(- phi0 - v_r_1(4))*-v_r_1(2) + cos(- phi0 - v_r_1(4))*(dzi + v_r_1(3)));
  edge_marker.points.push_back(p);

  p.x = -( cos(- phi0 - v_r_2(4))*-v_r_2(2) + sin(- phi0 - v_r_2(4))*(dzi + v_r_2(3)));
  p.y = a/2;
  p.z = -(-sin(- phi0 - v_r_2(4))*-v_r_2(2) + cos(- phi0 - v_r_2(4))*(dzi + v_r_2(3)));
  edge_marker.points.push_back(p);

  pub_edge_marker.publish(edge_marker);
  edge_marker.points.clear();
}